File: go/500.600/530.MinimumAbsoluteDifferenceInBst.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getMinimumDifference(root *TreeNode) int {
	var pre *TreeNode
	ans := math.MaxInt
	var dfs func(root *TreeNode)
	dfs = func(root *TreeNode) {
		if root.Left != nil {
			dfs(root.Left)
		}
		if pre != nil {
			ans = min(ans, root.Val-pre.Val)
		}
		pre = root
		if root.Right != nil {
			dfs(root.Right)
		}
	}
	dfs(root)
	return ans
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

func main() {
	root := &TreeNode{Val: 4}
	root.Left = &TreeNode{Val: 2}
	root.Left.Left = &TreeNode{Val: 1}
	root.Left.Right = &TreeNode{Val: 3}
	root.Right = &TreeNode{Val: 6}
	fmt.Println(getMinimumDifference(root), 1)

	root = &TreeNode{Val: 1}
	root.Left = &TreeNode{Val: 0}
	root.Right = &TreeNode{Val: 48}
	root.Right.Left = &TreeNode{Val: 12}
	root.Right.Right = &TreeNode{Val: 49}
	fmt.Println(getMinimumDifference(root), 1)
}

File: 500.600/530.MinimumAbsoluteDifferenceInBst.cpp
#include <iostream>
#include <algorithm>
using namespace std;

struct TreeNode
{
    int val{};
    TreeNode *left{};
    TreeNode *right{};
    TreeNode() {}
    TreeNode(int x) : val(x) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class MinimumAbsoluteDifferenceInBst
{
public:
    int getMinimumDifference(TreeNode *root)
    {
        ans = INT_MAX;
        last = -10e5;
        dfs(root);
        return ans;
    }

private:
    void dfs(TreeNode *root)
    {
        if (root->left)
        {
            dfs(root->left);
        }
        ans = min(ans, root->val - last);
        last = root->val;
        if (root->right)
        {
            dfs(root->right);
        }
    }

private:
    int ans{};
    int last{};
};

void del(TreeNode *root)
{
    if (root->left)
        del(root->left);
    if (root->right)
        del(root->right);
    delete root;
}

int main()
{
    MinimumAbsoluteDifferenceInBst cls;
    TreeNode *root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right = new TreeNode(6);
    cout << cls.getMinimumDifference(root) << " -> 1" << endl;
    del(root);

    root = new TreeNode(1);
    root->left = new TreeNode(0);
    root->right = new TreeNode(48);
    root->right->left = new TreeNode(12);
    root->right->right = new TreeNode(49);
    cout << cls.getMinimumDifference(root) << " -> 1" << endl;
    del(root);
    return 0;
}
